// nest-crud-typeorm-client/esm2015/lib/util/checks.util.js
import { objKeys } from './obj.util';
export const isUndefined = (val) => typeof val === 'undefined';
export const isNull = (val) => val === null;
export const isNil = (val) => isUndefined(val) || isNull(val);
export const isString = (val) => typeof val === 'string';
export const hasLength = (val) => val.length > 0;
export const isStringFull = (val) => isString(val) && hasLength(val);
export const isArrayFull = (val) => Array.isArray(val) && hasLength(val);
export const isArrayStrings = (val) => isArrayFull(val) && val.every((v) => isStringFull(v));
export const isObject = (val) => typeof val === 'object' && !isNull(val);
export const isObjectFull = (val) => isObject(val) && hasLength(objKeys(val));
export const isNumber = (val) => typeof val === 'number' && !Number.isNaN(val) && Number.isFinite(val);
export const isEqual = (val, eq) => val === eq;
export const isFalse = (val) => val === false;
export const isTrue = (val) => val === true;
export const isIn = (val, arr = []) => arr.some((o) => isEqual(val, o));
export const isBoolean = (val) => typeof val === 'boolean';
export const isNumeric = (val) => /^[+-]?([0-9]*[.])?[0-9]+$/.test(val);
export const isDateString = (val) => isStringFull(val) &&
    /^\d{4}-[01]\d-[0-3]\d(?:T[0-2]\d:[0-5]\d:[0-5]\d(?:\.\d+)?(?:Z|[-+][0-2]\d(?::?[0-5]\d)?)?)?$/g.test(val);
export const isDate = (val) => val instanceof Date;
export const isValue = (val) => isStringFull(val) || isNumber(val) || isBoolean(val) || isDate(val);
export const hasValue = (val) => isArrayFull(val) ? val.every((o) => isValue(o)) : isValue(val);
export const isFunction = (val) => typeof val === 'function';
