// ditagis-auth/esm2015/lib/services/dtg-log.service.js
import { Injectable, Inject } from '@angular/core';
import { DTGAUTH_CONFIG } from '../dtg-auth.config';
import { HttpClient } from '@angular/common/http';
import { DTGAuthService } from '../ditagis-auth.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../ditagis-auth.service";
import * as i3 from "../dtg-auth.config";
export class DTGLogService {
    constructor(http, authService, config) {
        this.http = http;
        this.authService = authService;
        this.config = config;
    }
    create(entity) {
        const url = this.config.logUrl || `${this.config.apiUrl}/sys/logger`;
        return this.http.post(url, JSON.stringify(entity), {
            headers: {
                'Content-Type': 'application/json',
                Authorization: 'Bearer ' + this.authService.getToken()
            }
        });
    }
}
DTGLogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGLogService_Factory() { return new DTGLogService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.DTGAuthService), i0.ɵɵinject(i3.DTGAUTH_CONFIG)); }, token: DTGLogService, providedIn: "root" });
DTGLogService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
DTGLogService.ctorParameters = () => [
    { type: HttpClient },
    { type: DTGAuthService },
    { type: undefined, decorators: [{ type: Inject, args: [DTGAUTH_CONFIG,] }] }
];
