// ditagis-auth/esm2015/lib/services/dtg-layerinfo.service.js
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { DTGAUTH_CONFIG } from '../dtg-auth.config';
import { DTGAuthService } from '../ditagis-auth.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../ditagis-auth.service";
import * as i3 from "../dtg-auth.config";
export class DTGLayerInfoService {
    constructor(http, authService, config) {
        this.http = http;
        this.authService = authService;
        this.config = config;
    }
    get() {
        const url = this.config.layerInfoUrl ||
            `${this.config.apiUrl}/auth/layerinfos`;
        return this.http.get(url, {
            headers: {
                Authorization: 'Bearer ' + this.authService.getToken()
            }
        });
    }
    getGuest() {
        const url = this.config.layerInfoUrl ||
            `${this.config.apiUrl}/auth/layerinfos/anonymous`;
        return this.http.get(url);
    }
}
DTGLayerInfoService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGLayerInfoService_Factory() { return new DTGLayerInfoService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.DTGAuthService), i0.ɵɵinject(i3.DTGAUTH_CONFIG)); }, token: DTGLayerInfoService, providedIn: "root" });
DTGLayerInfoService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
DTGLayerInfoService.ctorParameters = () => [
    { type: HttpClient },
    { type: DTGAuthService },
    { type: undefined, decorators: [{ type: Inject, args: [DTGAUTH_CONFIG,] }] }
];
