// nest-crud-typeorm-client/esm5/lib/util/checks.util.js
import { objKeys } from './obj.util';
export var isUndefined = function (val) { return typeof val === 'undefined'; };
export var isNull = function (val) { return val === null; };
export var isNil = function (val) { return isUndefined(val) || isNull(val); };
export var isString = function (val) { return typeof val === 'string'; };
export var hasLength = function (val) { return val.length > 0; };
export var isStringFull = function (val) { return isString(val) && hasLength(val); };
export var isArrayFull = function (val) { return Array.isArray(val) && hasLength(val); };
export var isArrayStrings = function (val) {
    return isArrayFull(val) && val.every(function (v) { return isStringFull(v); });
};
export var isObject = function (val) { return typeof val === 'object' && !isNull(val); };
export var isObjectFull = function (val) { return isObject(val) && hasLength(objKeys(val)); };
export var isNumber = function (val) {
    return typeof val === 'number' && !Number.isNaN(val) && Number.isFinite(val);
};
export var isEqual = function (val, eq) { return val === eq; };
export var isFalse = function (val) { return val === false; };
export var isTrue = function (val) { return val === true; };
export var isIn = function (val, arr) {
    if (arr === void 0) { arr = []; }
    return arr.some(function (o) { return isEqual(val, o); });
};
export var isBoolean = function (val) { return typeof val === 'boolean'; };
export var isNumeric = function (val) { return /^[+-]?([0-9]*[.])?[0-9]+$/.test(val); };
export var isDateString = function (val) {
    return isStringFull(val) &&
        /^\d{4}-[01]\d-[0-3]\d(?:T[0-2]\d:[0-5]\d:[0-5]\d(?:\.\d+)?(?:Z|[-+][0-2]\d(?::?[0-5]\d)?)?)?$/g.test(val);
};
export var isDate = function (val) { return val instanceof Date; };
export var isValue = function (val) {
    return isStringFull(val) || isNumber(val) || isBoolean(val) || isDate(val);
};
export var hasValue = function (val) {
    return isArrayFull(val) ? val.every(function (o) { return isValue(o); }) : isValue(val);
};
export var isFunction = function (val) { return typeof val === 'function'; };
