// ditagis-auth/bundles/ditagis-auth.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common/http'), require('@angular/router'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('ditagis-auth', ['exports', '@angular/core', '@angular/common/http', '@angular/router', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ditagis-auth'] = {}, global.ng.core, global.ng.common.http, global.ng.router, global.rxjs, global.rxjs.operators));
}(this, (function (exports, i0, i1, i2, rxjs, operators) { 'use strict';

    var DTGAUTH_CONFIG = new i0.InjectionToken('DTGAuthModuleOptions');

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    function __createBinding(o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    }
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                exports[p] = m[p];
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    result[k] = mod[k];
        result.default = mod;
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var TOKEN = 'token';
    var USER = 'user';
    var DTGAuthService = /** @class */ (function () {
        function DTGAuthService(http, config) {
            this.http = http;
            this.config = config;
        }
        DTGAuthService.prototype.authenticate = function (params) {
            return __awaiter(this, void 0, void 0, function () {
                var username, password, appId, url, result;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            username = params.username, password = params.password, appId = params.appId;
                            url = this.config.loginUrl || this.config.apiUrl + '/auth/login';
                            return [4 /*yield*/, this.http.post(url, {
                                    username: username, password: password,
                                    appId: appId
                                }).toPromise()];
                        case 1:
                            result = _a.sent();
                            if (result) {
                                localStorage.setItem(TOKEN, result.accessToken);
                                localStorage.setItem(USER, JSON.stringify(result));
                            }
                            return [2 /*return*/, result];
                    }
                });
            });
        };
        DTGAuthService.prototype.isAuthenticate = function () {
            return Boolean(this.getToken());
        };
        DTGAuthService.prototype.getToken = function () {
            return localStorage.getItem(TOKEN);
        };
        DTGAuthService.prototype.getInfo = function () {
            var user = localStorage.getItem(USER);
            if (user) {
                return JSON.parse(user);
            }
            return null;
        };
        DTGAuthService.prototype.deAuthenticate = function () {
            localStorage.removeItem(TOKEN);
            localStorage.removeItem(USER);
        };
        DTGAuthService.prototype.isAccess = function (params) {
            var appId = params.appId;
            var url = this.config.isAccessUrl ||
                this.config.apiUrl + "/auth/isaccess/" + appId;
            return this.http.get(url, {
                headers: {
                    Authorization: 'Bearer ' + this.getToken()
                }
            });
        };
        DTGAuthService.prototype.getAppInfo = function (params) {
            // Get a token from api server using the fetch api
            var appId = params.appId, isGuest = params.isGuest;
            var url = this.config.appInfoUrl || this.config.apiUrl + "/auth/appinfo/" + appId;
            var headers = new i1.HttpHeaders();
            if (!isGuest && this.isAuthenticate()) {
                headers = headers.append('Authorization', 'Bearer ' + this.getToken());
            }
            else {
                url = url + '/anonymous';
            }
            return this.http.get(url, {
                headers: headers
            });
        };
        return DTGAuthService;
    }());
    DTGAuthService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGAuthService_Factory() { return new DTGAuthService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(DTGAUTH_CONFIG)); }, token: DTGAuthService, providedIn: "root" });
    DTGAuthService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    DTGAuthService.ctorParameters = function () { return [
        { type: i1.HttpClient },
        { type: undefined, decorators: [{ type: i0.Inject, args: [DTGAUTH_CONFIG,] }] }
    ]; };

    var DTGLogService = /** @class */ (function () {
        function DTGLogService(http, authService, config) {
            this.http = http;
            this.authService = authService;
            this.config = config;
        }
        DTGLogService.prototype.create = function (entity) {
            var url = this.config.logUrl || this.config.apiUrl + "/sys/logger";
            return this.http.post(url, JSON.stringify(entity), {
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: 'Bearer ' + this.authService.getToken()
                }
            });
        };
        return DTGLogService;
    }());
    DTGLogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGLogService_Factory() { return new DTGLogService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(DTGAuthService), i0.ɵɵinject(DTGAUTH_CONFIG)); }, token: DTGLogService, providedIn: "root" });
    DTGLogService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    DTGLogService.ctorParameters = function () { return [
        { type: i1.HttpClient },
        { type: DTGAuthService },
        { type: undefined, decorators: [{ type: i0.Inject, args: [DTGAUTH_CONFIG,] }] }
    ]; };

    var DTGLayerInfoService = /** @class */ (function () {
        function DTGLayerInfoService(http, authService, config) {
            this.http = http;
            this.authService = authService;
            this.config = config;
        }
        DTGLayerInfoService.prototype.get = function () {
            var url = this.config.layerInfoUrl ||
                this.config.apiUrl + "/auth/layerinfos";
            return this.http.get(url, {
                headers: {
                    Authorization: 'Bearer ' + this.authService.getToken()
                }
            });
        };
        DTGLayerInfoService.prototype.getGuest = function () {
            var url = this.config.layerInfoUrl ||
                this.config.apiUrl + "/auth/layerinfos/anonymous";
            return this.http.get(url);
        };
        return DTGLayerInfoService;
    }());
    DTGLayerInfoService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGLayerInfoService_Factory() { return new DTGLayerInfoService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(DTGAuthService), i0.ɵɵinject(DTGAUTH_CONFIG)); }, token: DTGLayerInfoService, providedIn: "root" });
    DTGLayerInfoService.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    DTGLayerInfoService.ctorParameters = function () { return [
        { type: i1.HttpClient },
        { type: DTGAuthService },
        { type: undefined, decorators: [{ type: i0.Inject, args: [DTGAUTH_CONFIG,] }] }
    ]; };

    var DTGSimpleGuard = /** @class */ (function () {
        function DTGSimpleGuard(authService, router, config) {
            this.authService = authService;
            this.router = router;
            this.config = config;
        }
        DTGSimpleGuard.prototype.canActivate = function (route, state) {
            var isAuthenticate = this.authService.isAuthenticate();
            if (isAuthenticate) {
                // logged in so return true
                if (this.config.guard && this.config.guard.simple) {
                    if (this.config.guard.simple.canActive) {
                        this.config.guard.simple.canActive(route, state);
                    }
                }
                return true;
            }
            // not logged in so redirect to login page with the return url
            if (this.config.guard && this.config.guard.simple) {
                if (this.config.guard.simple.canNotActive) {
                    this.config.guard.simple.canNotActive(route, state);
                    return false;
                }
            }
            this.router.navigateByUrl(this.config.guard &&
                this.config.guard.simple && this.config.guard.simple.defaultNavigateUrl ?
                this.config.guard.simple.defaultNavigateUrl :
                '/auth', { queryParams: { returnUrl: state.url } });
            return false;
        };
        return DTGSimpleGuard;
    }());
    DTGSimpleGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGSimpleGuard_Factory() { return new DTGSimpleGuard(i0.ɵɵinject(DTGAuthService), i0.ɵɵinject(i2.Router), i0.ɵɵinject(DTGAUTH_CONFIG)); }, token: DTGSimpleGuard, providedIn: "root" });
    DTGSimpleGuard.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root' },] }
    ];
    DTGSimpleGuard.ctorParameters = function () { return [
        { type: DTGAuthService },
        { type: i2.Router },
        { type: undefined, decorators: [{ type: i0.Inject, args: [DTGAUTH_CONFIG,] }] }
    ]; };

    var DitagisAuthModule = /** @class */ (function () {
        function DitagisAuthModule() {
        }
        DitagisAuthModule.forRoot = function (options) {
            return {
                ngModule: DitagisAuthModule,
                providers: [
                    { provide: DTGAUTH_CONFIG, useValue: options },
                    DTGAuthService, DTGLogService, DTGLayerInfoService, DTGSimpleGuard
                ]
            };
        };
        return DitagisAuthModule;
    }());
    DitagisAuthModule.decorators = [
        { type: i0.NgModule, args: [{
                    declarations: [],
                    imports: [
                        i1.HttpClientModule
                    ]
                },] }
    ];

    (function (ApplicationEnvironmentEnum) {
        ApplicationEnvironmentEnum["Desktop"] = "D";
        ApplicationEnvironmentEnum["Web"] = "W";
        ApplicationEnvironmentEnum["Mobile"] = "M";
    })(exports.ApplicationEnvironmentEnum || (exports.ApplicationEnvironmentEnum = {}));

    var ResponseLogin = /** @class */ (function () {
        function ResponseLogin() {
        }
        return ResponseLogin;
    }());

    (function (ColumnTypeEnum) {
        ColumnTypeEnum["string"] = "esriFieldTypeString";
        ColumnTypeEnum["smallInteger"] = "esriFieldTypeSmallInteger";
        ColumnTypeEnum["integer"] = "esriFieldTypeInteger";
        ColumnTypeEnum["double"] = "esriFieldTypeDouble";
        ColumnTypeEnum["long"] = "esriFieldTypeLong";
        ColumnTypeEnum["date"] = "esriFieldTypeDate";
        ColumnTypeEnum["oid"] = "esriFieldTypeOID";
        ColumnTypeEnum["geometry"] = "esriFieldTypeGeometry";
    })(exports.ColumnTypeEnum || (exports.ColumnTypeEnum = {}));

    (function (LoggerActionTypeEnum) {
        LoggerActionTypeEnum["INSERT"] = "INSERT";
        LoggerActionTypeEnum["LOGIN"] = "LOGIN";
        LoggerActionTypeEnum["LOGOUT"] = "LOGOUT";
        LoggerActionTypeEnum["REMOVE"] = "REMOVE";
        LoggerActionTypeEnum["UPDATE"] = "UPDATE";
    })(exports.LoggerActionTypeEnum || (exports.LoggerActionTypeEnum = {}));

    (function (UserStatusEnum) {
        UserStatusEnum["Open"] = "OPEN";
        UserStatusEnum["Locked"] = "LOCKED";
        UserStatusEnum["Request"] = "REQUEST";
        UserStatusEnum["Expired"] = "EXPIRED";
    })(exports.UserStatusEnum || (exports.UserStatusEnum = {}));
    (function (RoleStatusEnum) {
        RoleStatusEnum["Open"] = "OPEN";
        RoleStatusEnum["Lock"] = "LOCK";
        RoleStatusEnum["Hide"] = "HIDE";
    })(exports.RoleStatusEnum || (exports.RoleStatusEnum = {}));

    var ErrorInterceptor = /** @class */ (function () {
        function ErrorInterceptor(authService, config) {
            this.authService = authService;
            this.config = config;
        }
        ErrorInterceptor.prototype.intercept = function (request, next) {
            var _this = this;
            return next.handle(request).pipe(operators.catchError(function (err) {
                var url = _this.config && _this.config.loginUrl || 'login';
                if (request.url.search(url) === -1) {
                    if (err.status === 401) {
                        // auto logout if 401 response returned from api
                        _this.authService.deAuthenticate();
                        location.reload(true);
                    }
                }
                var error = err.error.message || err.statusText || 'Có lỗi xảy ra';
                return rxjs.throwError(error);
            }));
        };
        return ErrorInterceptor;
    }());
    ErrorInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function ErrorInterceptor_Factory() { return new ErrorInterceptor(i0.ɵɵinject(DTGAuthService), i0.ɵɵinject(DTGAUTH_CONFIG)); }, token: ErrorInterceptor, providedIn: "root" });
    ErrorInterceptor.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root', deps: [DTGAuthService] },] }
    ];
    ErrorInterceptor.ctorParameters = function () { return [
        { type: DTGAuthService },
        { type: undefined, decorators: [{ type: i0.Inject, args: [DTGAUTH_CONFIG,] }] }
    ]; };

    var JwtInterceptor = /** @class */ (function () {
        function JwtInterceptor(authService) {
            this.authService = authService;
        }
        JwtInterceptor.prototype.intercept = function (request, next) {
            // add authorization header with jwt token if available
            var isAuthenticate = this.authService.isAuthenticate();
            if (isAuthenticate) {
                var token = this.authService.getToken();
                request = request.clone({
                    setHeaders: {
                        Authorization: "Bearer " + token
                    }
                });
            }
            return next.handle(request);
        };
        return JwtInterceptor;
    }());
    JwtInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function JwtInterceptor_Factory() { return new JwtInterceptor(i0.ɵɵinject(DTGAuthService)); }, token: JwtInterceptor, providedIn: "root" });
    JwtInterceptor.decorators = [
        { type: i0.Injectable, args: [{ providedIn: 'root', deps: [DTGAuthService] },] }
    ];
    JwtInterceptor.ctorParameters = function () { return [
        { type: DTGAuthService }
    ]; };

    /*
     * Public API Surface of ditagis-auth
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.DTGAUTH_CONFIG = DTGAUTH_CONFIG;
    exports.DTGAuthService = DTGAuthService;
    exports.DTGLayerInfoService = DTGLayerInfoService;
    exports.DTGLogService = DTGLogService;
    exports.DTGSimpleGuard = DTGSimpleGuard;
    exports.DitagisAuthModule = DitagisAuthModule;
    exports.ErrorInterceptor = ErrorInterceptor;
    exports.JwtInterceptor = JwtInterceptor;
    exports.ResponseLogin = ResponseLogin;
    exports.ɵa = DTGLogService;
    exports.ɵb = DTGLayerInfoService;
    exports.ɵc = DTGSimpleGuard;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
