// ditagis-auth/esm2015/lib/guards/simple.guard.js
import { Router } from '@angular/router';
import { Injectable, Inject } from '@angular/core';
import { DTGAuthService } from '../ditagis-auth.service';
import { DTGAUTH_CONFIG } from '../dtg-auth.config';
import * as i0 from "@angular/core";
import * as i1 from "../ditagis-auth.service";
import * as i2 from "@angular/router";
import * as i3 from "../dtg-auth.config";
export class DTGSimpleGuard {
    constructor(authService, router, config) {
        this.authService = authService;
        this.router = router;
        this.config = config;
    }
    canActivate(route, state) {
        const isAuthenticate = this.authService.isAuthenticate();
        if (isAuthenticate) {
            // logged in so return true
            if (this.config.guard && this.config.guard.simple) {
                if (this.config.guard.simple.canActive) {
                    this.config.guard.simple.canActive(route, state);
                }
            }
            return true;
        }
        // not logged in so redirect to login page with the return url
        if (this.config.guard && this.config.guard.simple) {
            if (this.config.guard.simple.canNotActive) {
                this.config.guard.simple.canNotActive(route, state);
                return false;
            }
        }
        this.router.navigateByUrl(this.config.guard &&
            this.config.guard.simple && this.config.guard.simple.defaultNavigateUrl ?
            this.config.guard.simple.defaultNavigateUrl :
            '/auth', { queryParams: { returnUrl: state.url } });
        return false;
    }
}
DTGSimpleGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGSimpleGuard_Factory() { return new DTGSimpleGuard(i0.ɵɵinject(i1.DTGAuthService), i0.ɵɵinject(i2.Router), i0.ɵɵinject(i3.DTGAUTH_CONFIG)); }, token: DTGSimpleGuard, providedIn: "root" });
DTGSimpleGuard.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
DTGSimpleGuard.ctorParameters = () => [
    { type: DTGAuthService },
    { type: Router },
    { type: undefined, decorators: [{ type: Inject, args: [DTGAUTH_CONFIG,] }] }
];
