// ditagis-auth/esm2015/lib/ditagis-auth.service.js
import { __awaiter } from "tslib";
import { Injectable, Inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { DTGAUTH_CONFIG } from './dtg-auth.config';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./dtg-auth.config";
const TOKEN = 'token';
const USER = 'user';
export class DTGAuthService {
    constructor(http, config) {
        this.http = http;
        this.config = config;
    }
    authenticate(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { username, password, appId } = params;
            const url = this.config.loginUrl || this.config.apiUrl + '/auth/login';
            const result = yield this.http.post(url, {
                username, password,
                appId
            }).toPromise();
            if (result) {
                localStorage.setItem(TOKEN, result.accessToken);
                localStorage.setItem(USER, JSON.stringify(result));
            }
            return result;
        });
    }
    isAuthenticate() {
        return Boolean(this.getToken());
    }
    getToken() {
        return localStorage.getItem(TOKEN);
    }
    getInfo() {
        const user = localStorage.getItem(USER);
        if (user) {
            return JSON.parse(user);
        }
        return null;
    }
    deAuthenticate() {
        localStorage.removeItem(TOKEN);
        localStorage.removeItem(USER);
    }
    isAccess(params) {
        const { appId } = params;
        const url = this.config.isAccessUrl ||
            `${this.config.apiUrl}/auth/isaccess/${appId}`;
        return this.http.get(url, {
            headers: {
                Authorization: 'Bearer ' + this.getToken()
            }
        });
    }
    getAppInfo(params) {
        // Get a token from api server using the fetch api
        const { appId, isGuest } = params;
        let url = this.config.appInfoUrl || `${this.config.apiUrl}/auth/appinfo/${appId}`;
        let headers = new HttpHeaders();
        if (!isGuest && this.isAuthenticate()) {
            headers = headers.append('Authorization', 'Bearer ' + this.getToken());
        }
        else {
            url = url + '/anonymous';
        }
        return this.http.get(url, {
            headers
        });
    }
}
DTGAuthService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DTGAuthService_Factory() { return new DTGAuthService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.DTGAUTH_CONFIG)); }, token: DTGAuthService, providedIn: "root" });
DTGAuthService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
DTGAuthService.ctorParameters = () => [
    { type: HttpClient },
    { type: undefined, decorators: [{ type: Inject, args: [DTGAUTH_CONFIG,] }] }
];
