// ditagis-auth/esm2015/lib/interceptors/jwt.interceptor.js
import { Injectable } from '@angular/core';
import { DTGAuthService } from '../ditagis-auth.service';
import * as i0 from "@angular/core";
import * as i1 from "../ditagis-auth.service";
export class JwtInterceptor {
    constructor(authService) {
        this.authService = authService;
    }
    intercept(request, next) {
        // add authorization header with jwt token if available
        const isAuthenticate = this.authService.isAuthenticate();
        if (isAuthenticate) {
            const token = this.authService.getToken();
            request = request.clone({
                setHeaders: {
                    Authorization: `Bearer ${token}`
                }
            });
        }
        return next.handle(request);
    }
}
JwtInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function JwtInterceptor_Factory() { return new JwtInterceptor(i0.ɵɵinject(i1.DTGAuthService)); }, token: JwtInterceptor, providedIn: "root" });
JwtInterceptor.decorators = [
    { type: Injectable, args: [{ providedIn: 'root', deps: [DTGAuthService] },] }
];
JwtInterceptor.ctorParameters = () => [
    { type: DTGAuthService }
];
